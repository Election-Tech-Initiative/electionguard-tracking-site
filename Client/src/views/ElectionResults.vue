<template>
  <PreElectionGuard>
    <div class="section">
      <div class="container">
        <div class="content">
        </div>
      </div>
    </div>

    <LoadingIndicator v-if="isLoading" />
    <div class="section results" v-else-if="showElectionResults">
      <div class="container">
        <div class="content">
          <ElectionResultsTally :electionResults="electionResults" />
        </div>
      </div>
    </div>
  </PreElectionGuard>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import { ElectionSummary } from '@/models/ElectionSummary';
import { BallotEntry } from '@/models/BallotEntry';
import { Tally } from '@/models/Tally';

import ElectionResultsTally from '@/components/ElectionResultsTally.vue';
import LoadingIndicator from '@/components/LoadingIndicator.vue';
import PreElectionGuard from '@/components/PreElectionGuard.vue';

@Component({
  components: {
    ElectionResultsTally,
    LoadingIndicator,
    PreElectionGuard,
  },
})
export default class ElectionResults extends Vue {
  public isLoading: boolean = true;

  // This has to be initialized in order for it to be reactive
  public electionResults: ElectionSummary = {
    isComplete: false,
    ballotEntries: []
  };

  constructor() {
    super();

    this.queryResults();
  }

  public get showElectionResults(): boolean {
    return !!this.electionResults && this.electionResults.isComplete;
  }

  public async queryResults(): Promise<void> {
    this.isLoading = true;
    try {
      const url = `${this.appSettings.apiUrl}/summary`;
      const response = await this.$http.get(url);
      if (response.status === 200) {
        this.electionResults = response.data;
      } else {
        console.log(`Error: ${response.statusText}`);
      }
    } catch (error) {
      console.log(`Error while getting election results summary`);
      console.log(error);
    }
    this.isLoading = false;
  }
}
</script>

<style lang="scss" scoped>
.section.results {
  background-color: $white;
}
</style>
