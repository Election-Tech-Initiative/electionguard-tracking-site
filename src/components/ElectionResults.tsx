import React from 'react';
import { useElectionResults } from '../data/queries';
import { Election } from '../models/election';
import { ElectionResultsSummary } from '../models/tally';
import AsyncContent from './AsyncContent';
import ContestResults from './ContestResults';
import LargeCard from './LargeCard';

const errorMessage = 'Unable to retrieve election results at this time.';

export interface ElectionResultsProps {
    election: Election;
}

/**
 * Render the results of the election
 */
const ElectionResults: React.FunctionComponent<ElectionResultsProps> = ({ election }) => {
    const electionResultsQuery = useElectionResults(election.id);

    return (
        <AsyncContent query={electionResultsQuery} errorMessage={errorMessage}>
            {(results) => {
                const contests = getContests(election, results);
                return (
                    <>
                        {contests.map((contest) => (
                            <LargeCard key={contest.id}>
                                <ContestResults
                                    results={contest.results}
                                    contest={contest.description!}
                                    candidates={election.election_description.candidates}
                                />
                            </LargeCard>
                        ))}
                    </>
                );
            }}
        </AsyncContent>
    );
};

function getContests(election: Election, results: ElectionResultsSummary) {
    const contests = Object.entries(results.election_results)
        .map(([contestId, contestResults]) => ({
            id: contestId,
            results: contestResults,
            description: election.election_description.contests.find((c) => c.object_id === contestId),
        }))
        .filter((c) => !!c.description);
    return contests;
}

export default ElectionResults;
