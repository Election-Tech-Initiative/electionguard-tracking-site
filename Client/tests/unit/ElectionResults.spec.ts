import { mount } from '@vue/test-utils';

import ElectionResults from '@/views/ElectionResults.vue';

jest.mock('@/components/ElectionResultsTally.vue', () => ({ default: jest.fn() }));
jest.mock('@/components/LoadingIndicator.vue', () => ({ default: jest.fn() }));
jest.mock('@/components/PreElectionGuard.vue', () => ({ default: jest.fn() }));

describe('ElectionResults.vue', () => {
    const getMountedWrapper = () => mount(ElectionResults, {
        mocks: {
            appSettings: {
                apiUrl: 'mocked.url',
            },
            $http: {
                get: () => Promise.resolve({ data: { isComplete: true, ballotEntries: [] }, status: 200 }),
            },
        },
        stubs: [
            'ElectionResultsTally',
            'LoadingIndicator',
            'PreElectionGuard',
        ],
    });

    it('renders with LoadingIndicator', () => {
        const wrapper = getMountedWrapper();
        expect(wrapper.element).toMatchSnapshot('withLoadingIndicator');
    });

    it('renders with ElectionResultsTally after async fetch', async () => {
        const wrapper = getMountedWrapper();
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot('withElectionResultsTally');
    });
});
