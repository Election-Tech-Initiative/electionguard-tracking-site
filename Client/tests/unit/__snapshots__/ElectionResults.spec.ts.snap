// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectionResults.vue renders with ElectionResultsTally after async fetch: withElectionResultsTally 1`] = `
<preelectionguard-stub>
  <div
    class="section"
  >
    <div
      class="container"
    >
      <div
        class="content"
      />
    </div>
  </div>
   
  <div
    class="section results"
  >
    <div
      class="container"
    >
      <div
        class="content"
      >
        <electionresultstally-stub
          electionresults="[object Object]"
        />
      </div>
    </div>
  </div>
</preelectionguard-stub>
`;

exports[`ElectionResults.vue renders with LoadingIndicator: withLoadingIndicator 1`] = `
<preelectionguard-stub>
  <div
    class="section"
  >
    <div
      class="container"
    >
      <div
        class="content"
      />
    </div>
  </div>
   
  <loadingindicator-stub />
</preelectionguard-stub>
`;
